server/src/main/java/com/jinyframework/keva/server/core/WriteLog.java

```diff
@@ -0,0 +1,115 @@
+package com.jinyframework.keva.server.core;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A circular array that supports multiple thread reads and single thread write
+ */
+@Slf4j
+public class WriteLog {
+    private final ReentrantLock lock = new ReentrantLock();
+
+
+    private final AtomicInteger currentOffset = new AtomicInteger(0); // increase every write
+
+    private final AtomicInteger startingOffset = new AtomicInteger(0); // denotes offset at index 0
+
+    private final int capacity;
+    private final ArrayList<String> buffer;
+
+    /**
+     * Instantiates a new Command log.
+     *
+     * @param capacity the capacity
+     */
+    public WriteLog(int capacity) {
+        this.capacity = capacity;
+        buffer = new ArrayList<>(Arrays.asList(new String[capacity]));
+    }
+
+    /**
+     * Buffer the command, if capacity limit is reached, it will wraps around and override data in the first index
+     *
+     * @param command the command
+     */
+    public void buffer(String command) {
+        final int offset = currentOffset.getAndIncrement();
+        final int index = offset % capacity;
+        final int startOffset = startingOffset.get();
+        if (offset >= startOffset + capacity) {
+            startingOffset.set(offset);
```

Comment:
During this, if another thread running the same function, it can overwrite startingOffset, is it ok?
like a buffer with cap at 3, and 2 thread as followed:
- offset = 3 >= startingOffset + cap (0+3) --> write startingOffset to 3
- offset =4 >= startingOffset + cap (0+3) (it starting offset can still be 3) --> write startingOffset = 4

Reply:
Yeah, I'll revert to the old code, putting most of it inside a lock for now. Thanks for spotting